fog/ledger/server/src/block_service.rs

```diff
@@ -45,65 +41,46 @@ impl<L: Ledger + Clone> BlockService<L> {
     fn get_blocks_impl(&mut self, request: BlockRequest) -> Result<BlockResponse, RpcStatus> {
         mc_common::trace_time!(self.logger, "Get Blocks");
 
-        let mut result = BlockResponse::new();
+        let block_indices = request
+            .ranges
+            .iter()
+            .flat_map(|range| range.start_block..range.end_block)
+            .collect::<Vec<_>>();
 
-        result.num_blocks = self
-            .ledger
-            .num_blocks()
-            .map_err(|err| rpc_database_err(err, &self.logger))?;
-        result.global_txo_count = self
-            .ledger
-            .num_txos()
+        let BlocksDataResponse {
+            results,
+            latest_block,
+        } = self
+            .block_provider
+            .get_blocks_data(block_indices.as_slice())
             .map_err(|err| rpc_database_err(err, &self.logger))?;
 
-        for range in request.ranges.iter() {
-            for block_idx in range.start_block..range.end_block {
-                match self.get_block(block_idx) {
-                    Ok(block) => result.blocks.push(block),
-                    // TODO: signal internal error for some errors?
-                    Err(err) => {
-                        log::error!(self.logger, "DbError getting block {}: {}", block_idx, err)
-                    }
-                };
-            }
-        }
+        let mut response = BlockResponse::new();
+        response.num_blocks = latest_block.index + 1;
+        response.global_txo_count = latest_block.cumulative_txo_count;
```

Comment:
This kind of change is in a few places - moving from `ledger.num_blocks()` and `ledger.num_txos()` to getting this data from the actual last block. I think that while slightly less efficient (since we need to read and de-serialize the `Block`), the behavior is identical.